soc/xtensa/intel_adsp/tools/hw-service.py

```diff
@@ -0,0 +1,449 @@
+        self.is_load_fw = self.board.is_load_fw()
+
+    def set_dev_lock(self, user, passcode):
+        if not self.lock_enable:
+            return True
+
+        if not self.available:
+            return False
+
+        self.lock.acquire()
+        self.available = False
+        self.occpied_user = user
+        self.passcode = passcode
+        self.lock.release()
+        return True
+
+    def set_dev_unlock(self, user, passcode):
+        if self.lock_enable:
+            self.lock.acquire()
+            self.available = True
+            self.occpied_user = None
+            self.passcode = None
+            self.lock.release()
+
+    def is_dev_available(self, user):
+        if not self.lock_enable:
+            return True
+
+        if self.occpied_user == user or self.available:
+            return True
+        else:
+            return False
+
+    def set_fw_ready(self, fw_recv):
+        if fw_recv:
+            self.fw_queue.put(fw_recv)
+
+    def is_fw_ready(self):
+        self.fw_file = self.fw_queue.get()
+        log.info(f"Current FW is {self.fw_file}")
+
+        if self.fw_file:
+            if self.fw_file == "OTHER":
+                self.fw_file = None
+            return True
+        else:
+            return False
+
+    def cleanup(self):
+        self.lock.acquire()
+        self.script = None
+        if self.fw_file:
+            os.remove(self.fw_file)
+        self.fw_file = None
+        self.lock.release()
+
+    def get_script(self):
+        if os.geteuid() != 0:
+            self.script = ([f'sudo', f'{self.load_cmd}'])
+        else:
+            self.script = ([f'{self.load_cmd}'])
+
+        if self.is_load_fw:
+            self.script.append(f'{self.fw_file}')
+
+        if self.board.params:
+            for param in self.board.params:
+                self.script.append(param)
+
+        log.info(f'run script: {self.script}')
+        return self.script
+
+class board_config():
+    def __init__(self, board, serial, load_cmd):
+
+        board_intel_adsp = ['intel_adsp_cavs15', 'intel_adsp_cavs18',
+                            'intel_adsp_cavs20', 'intel_adsp_cavs20_jsl',
+                            'intel_adsp_cavs25', 'intel_adsp_cavs25_tgph']
+
+        board_x86_hypervisor = ['acrn_ehl_crb', 'acrn']
+
+        self.board = board          # the board name
+        self.serial = serial        # serial port using
+        self.load_cmd = load_cmd    # cmd for loading
+        self.params = []            # params of loading cmd
+        self.load_fw = False        # need to upload FW
+
+        if not self.board or self.board in board_intel_adsp:
+            log.info(f"Set board to intel_adsp by default.")
+            self.load_fw = True
+            if not self.load_cmd:
+                self.load_cmd = "./cavstool.py"
+
+        elif self.board in board_x86_hypervisor:
+            if not load_cmd:
+                self.load_cmd = "./acrn-test-pty.exp"
```

Comment:
This is not even in tree, is it ?

Reply:
> I like the separation but instead of copying the code in one commit and deleting in a later commit (which is not even the next commit), can you please MOVE as much code as possible in a single commit? This will help git track the moved lines of code and it will also make the code review and git log easier to follow.

Yes, I will fix this. Thanks!

> Please don't mix the addition of brand new features in the same commit that copies or moves a lot of code https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes Actually, for Continuous Integration reasons please separate these in different Pull Requests.

You are right. I should do it this way. I will keep changes of CAVS only in this PR, and support for other boards should be another PR.

> Instead of adding a lock (and try to track IP addresses?), why not just accept a single TCP connection at a time?
Add a lock is trying to prevent two users to run west/twister from accessing the board at the same time. Because we have the "request" and "log" service, these two services are occupied by a single user in a time period. I not sure if a single TCP connection can solve this. I will try it.

> I don't understand how separating the code into 2 different files can "fix" a ROM boot failure, this does not make sense to me sorry, can you elaborate? So far I've seen two totally different kinds of boot failures:

Sorry for making confusing here. The major difference is I try to spawn another process to run the pure cavstool.py, and if it is stuck, the sever can kill it and continue to run the next test. It's kind of like what the original cavstool.py and cavstwist.sh did.
Another simple fix is to add a timeout in the "waiting for ROM starting" while loop: 
```
    log.info(f"Wait for ROM startup, ADSPCS = 0x{dsp.ADSPCS:x}")
    time.sleep(1)
    while (dsp.SRAM_FW_STATUS >> 24) != 5: pass
```
Otherwise it would stuck here forever and block all following tests.
I was considering to add this earlier, but finally I thought the best way is to totally split the server and the fw_loader. It seems like I have added too much code in this cavstool.py and might confuse the debugging of the tool or fw_loader itself. How do you think?

> * Intermittent firmware or firmware authentication issues. These cannot be "fixed" by moving Python code around. Python code can at best perform retries or other workarounds but it cannot "solve" anything
> * Lack of delimiter and other misuse of raw TCP sockets in discussed in [#46864 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/46864#issuecomment-1167545791) and links. These _can_ be fixed in Python but not by shuffling code around files.
I will continue on improving this. Indeed your suggestion is right.

> Not a fan of the new filename `hw-service.py` but I don't have any great alternative to offer. Maybe `remote-fw-service.py`?
Good suggestion. It sounds better. I will update it.

Reply:
????

Reply:
Yes, remove it.